### PR TITLE
feat: encode FixedSizeBinary in JSON as hex string

### LINKED ISSUE #5620: Serialize `FixedSizeBinary` as HEX with JSON Writer
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

I would like to have `FixedSizeBinary` columns be serialized by the JSON writer as HEX, as they are for the CSV writer.

**Describe the solution you'd like**

Columns that are `FixedSizeBinary` be serialized as HEX strings. The CSV writer defaults to the default arrow-array formatter, which does this: https://github.com/apache/arrow-rs/blob/master/arrow-cast/src/display.rs#L746-L754

**Describe alternatives you've considered**

Alternatively, I could change my queries to potentially cast the column I am querying for, so this isn't a deal breaker for me, but I feel as though since the CSV and default formatters do this for me, the JSON should as well.

**Additional context**

The error returned originates from here: https://github.com/apache/arrow-rs/blob/master/arrow-json/src/writer/encoder.rs#L128

### PR DESCRIPTION
**Pull request by hiltontj**
# Which issue does this PR close?

Closes #5620.

# Rationale for this change
 

See #5620.

# What changes are included in this PR?

Adds encoding support to the JSON writer for the `FixedSizeBinary` data type.

A simple test was added as well. I can add more extensive testing if needed.

# Are there any user-facing changes?

Potentially some documentation to state that `FixedSizeBinary` is encoded as hex strings.

### CODE REVIEW

File: arrow-json/Cargo.toml [hunk 1]
```diff
@@ -56,6 +56,7 @@ tokio = { version = "1.27", default-features = false, features = ["io-util"] }
 bytes = "1.4"
 criterion = { version = "0.5", default-features = false }
 rand = { version = "0.8", default-features = false, features = ["std", "std_rng"] }
+hex = "0.4.3"
```
Comment (tustvold) [inline arrow-json/Cargo.toml:59]: Do we need this?

Comment (hiltontj) [reply arrow-json/Cargo.toml:59]: No, it is not required. The reason I added it was so that in the test I added, it was apparent that the bytes in the column resulted in the correct HEX string. I will see if I can write the test in a way that gets by without it.

**Comment by hiltontj** [comment]
I realize now, that the test I wrote is complete bogus (not valid JSON). So I am reworking that.

File: arrow-json/Cargo.toml (hunk 1 above)
Comment (hiltontj) [reply arrow-json/Cargo.toml]: I updated the test so that it no longer relies on the `hex` crate, and removed the dep. I think it would be nice to have a test that fully round-trips data, first by decoding from HEX to the `FixedSizeBinary`, then encoding back. But, I am not sure if we want to support this behaviour on the JSON reader, and this issue is only about the write side.

File: arrow-json/src/writer/encoder.rs [hunk 2]
```diff
@@ -443,3 +448,17 @@ impl<'a> Encoder for MapEncoder<'a> {
         out.push(b'}');
     }
 }
+
+struct FixedSizeBinaryEncoder(FixedSizeBinaryArray);
+
+impl Encoder for FixedSizeBinaryEncoder {
+    fn encode(&mut self, idx: usize, out: &mut Vec<u8>) {
+        let v = self.0.value(idx);
```
Comment (alamb) [inline arrow-json/src/writer/encoder.rs:456]: I think this needs to check `is_valid` before serializing (I think you'll see the need when testing for null)

**Review by alamb — COMMENTED** [review]
Thanks @hiltontj ! This is looking close -- I am not sure it handles nulls properly yet, but I think that should be straightforward to add

**Comment by hiltontj** [comment]
@alamb - I think I addressed your comments. With https://github.com/apache/arrow-rs/pull/5622/commits/35371d39d59b020a4da81422f024b6147067951a, explicit nulls were actually being handled correctly in the test, but I believe this was because of the wrapping list/map encoders. So, I extended the explicit null support down to the `FixedSizeBinaryEncoder` in https://github.com/apache/arrow-rs/pull/5622/commits/9768a43c96276008b342970d0254d23355faef57 directly, to be sure.

File: arrow-json/src/writer.rs [hunk 3]
```diff
@@ -2137,4 +2139,80 @@ mod tests {
...
+                writer.close().unwrap();
+                serde_json::from_slice(&buf).unwrap()
+            };
+
+            assert_eq!(
+                json!([
+                    {
+                        "bytes": "68656c6c6f20776f726c64"
+                    },
+                    {}, // empty because nulls are omitted
```
Comment (alamb) [inline arrow-json/src/writer.rs:2209]: 👍

**Review by alamb — APPROVED** [review]
Looks great to me -- thank you @hiltontj  for adding the new tests

**Comment by alamb** [comment]
I think this is good to go -- any concerns @tustvold ?

**Comment by hiltontj** [comment]
@alamb - I changed the `FixedSizeBinaryEncoder` to use a borrow of the `FixedSizeBinaryArray`, instead of cloning it.

**Comment by hiltontj** [comment]
I also noticed that the `StringEncoder` and `BooleanEncoder` were needlessly cloning, so I updated them to use borrows, and removed the `.clone()`s. Hope you don't mind.

File: arrow-json/src/writer/encoder.rs [hunk 4]
```diff
@@ -443,3 +448,33 @@ impl<'a> Encoder for MapEncoder<'a> {
...
+        Self {
+            array,
+            explicit_nulls: options.explicit_nulls,
+        }
+    }
+}
+
+impl<'a> Encoder for FixedSizeBinaryEncoder<'a> {
+    fn encode(&mut self, idx: usize, out: &mut Vec<u8>) {
+        if self.array.is_valid(idx) {
```
Comment (tustvold) [inline arrow-json/src/writer/encoder.rs:468]: IIRC nullability is handled by the parent and so this shouldn't be necessary

**Comment by tustvold** [comment]
> were needlessly cloning

A couple of reference counts are irrelevant compared to encoding overheads, but I suppose it can't hurt to avoid them

File: arrow-json/src/writer/encoder.rs (hunk 4 above)
Comment (alamb) [reply arrow-json/src/writer/encoder.rs:468]: I double checked by making the following change locally and the tests all still pass. Thus I agree the valid is not needed

```diff
diff --git a/arrow-json/src/writer/encoder.rs b/arrow-json/src/writer/encoder.rs
index 27fa5747288..470d2fadd06 100644
--- a/arrow-json/src/writer/encoder.rs
+++ b/arrow-json/src/writer/encoder.rs
@@ -101,7 +101,7 @@ fn make_encoder_impl<'a>(

         DataType::FixedSizeBinary(_) => {
             let array = array.as_any().downcast_ref::<FixedSizeBinaryArray>().unwrap();
-            (Box::new(FixedSizeBinaryEncoder::new(array, options)) as _, array.nulls().cloned())
+            (Box::new(FixedSizeBinaryEncoder::new(array)) as _, array.nulls().cloned())
         }

         DataType::Struct(fields) => {
@@ -451,14 +451,12 @@ impl<'a> Encoder for MapEncoder<'a> {

 struct FixedSizeBinaryEncoder<'a> {
     array: &'a FixedSizeBinaryArray,
-    explicit_nulls: bool,
 }

 impl<'a> FixedSizeBinaryEncoder<'a> {
-    fn new(array: &'a FixedSizeBinaryArray, options: &EncoderOptions) -> Self {
+    fn new(array: &'a FixedSizeBinaryArray) -> Self {
         Self {
             array,
-            explicit_nulls: options.explicit_nulls,
         }
     }
 }
@@ -473,8 +471,6 @@ impl<'a> Encoder for FixedSizeBinaryEncoder<'a> {
                 write!(out, "{byte:02x}").unwrap();
             }
             out.push(b'"');
-        } else if self.explicit_nulls {
-            out.extend_from_slice(b"null");
         }
     }
 }
```

**Review by alamb — APPROVED** [review]
Thanks @hiltontj  -- I think this PR could be merged as is.

@tustvold  has a good suggestion for simplificiation https://github.com/apache/arrow-rs/pull/5622/files#r1562191540 but I also think we could do that as a follow on PR if you prefer

Just let me know what you would prefer

File: arrow-json/src/writer/encoder.rs (hunk 4 above)
Comment (hiltontj) [reply arrow-json/src/writer/encoder.rs:468]: Indeed, as per https://github.com/apache/arrow-rs/pull/5622#issuecomment-2047812089, when I first added the test, it was passing without adding the `explicit_nulls` option to the `FixedSizeBinaryEncoder`, but I added it anyways. I am totally fine with taking it out.

**Comment by hiltontj** [comment]
@alamb - I will update as suggested on this PR and re-request review, no problem!

**Review by alamb — APPROVED** [review]
Thanks @hiltontj  for sticking with this. It looks very nice at the end 🙏
